arch/x86/x86_64/fsgsbase.c

```diff
@@ -0,0 +1,120 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2024, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <uk/arch/lcpu.h>
+#include <uk/ctors.h>
+#include <x86/cpu.h>
+
+void (*wrgsbasefn)(__uptr gsbase);
+__uptr (*rdgsbasefn)(void);
+
+static void wrmsrgsbase(__uptr gsbase)
+{
+	wrmsrl(X86_MSR_GS_BASE, gsbase);
+}
+
+static __uptr rdmsrgsbase(void)
+{
+	return rdmsrl(X86_MSR_GS_BASE);
+}
+
+static __uptr rdgsbase_cr4fsgsbase(void)
+{
+	__uptr gsbase;
+
+	__asm__ __volatile__(
+		"rdgsbase	%0"
+		: "=r" (gsbase)
+		:
+		: "memory"
+	);
+
+	return gsbase;
+}
+
+static void wrgsbase_cr4fsgsbase(__uptr gsbase)
+{
+	__asm__ __volatile__(
+		"wrgsbase	%0"
+		:
+		: "r" (gsbase)
+		: "memory"
+	);
+}
+
+void (*wrkgsbasefn)(__uptr kgsbase);
+__uptr (*rdkgsbasefn)(void);
+
+static void wrmsrkgsbase(__uptr kgsbase)
+{
+	wrmsrl(X86_MSR_KERNEL_GS_BASE, kgsbase);
+}
+
+static __uptr rdmsrkgsbase(void)
+{
+	return rdmsrl(X86_MSR_KERNEL_GS_BASE);
+}
+
+void (*wrfsbasefn)(__uptr fsbase);
+__uptr (*rdfsbasefn)(void);
+
+static void wrmsrfsbase(__uptr fsbase)
+{
+	wrmsrl(X86_MSR_FS_BASE, fsbase);
+}
+
+static __uptr rdmsrfsbase(void)
+{
+	return rdmsrl(X86_MSR_FS_BASE);
+}
+
+static __uptr rdfsbase_cr4fsgsbase(void)
+{
+	__uptr fsbase;
+
+	__asm__ __volatile__(
+		"rdfsbase	%0"
+		: "=r" (fsbase)
+		:
+		: "memory"
+	);
+
+	return fsbase;
+}
+
+static void wrfsbase_cr4fsgsbase(__uptr fsbase)
+{
+	__asm__ __volatile__(
+		"wrfsbase	%0"
+		:
+		: "r" (fsbase)
+		: "memory"
+	);
+}
+
+
```

Comment:
Hmm interesting how checkpatch did not see this empty line. Will remove it on next relevant force push, I don't want to unnecessarily trigger another CI run.